xonsh/environ.py

```diff
@@ -816,17 +819,25 @@ def _yield_executables(directory, name):
 
 
 def locate_binary(name):
-    if os.path.isfile(name) and name != os.path.basename(name):
-        return name
-    directories = builtins.__xonsh_env__.get('PATH')
-    # Windows users expect to be able to execute files in the same directory
-    # without `./`
+    """Locates an executable on the file system."""
     if ON_WINDOWS:
-        directories = [_get_cwd()] + list(directories)
-    try:
-        return next(chain.from_iterable(_yield_executables(directory, name) for
-                    directory in directories if os.path.isdir(directory)))
-    except StopIteration:
+        # Windows users expect to be able to execute files in the same
+        # directory without `./`
+        cwd = _get_cwd()
+        if os.path.isfile(name):
+            return os.path.abspath(os.path.relpath(name, cwd))
+        exts = builtins.__xonsh_env__['PATHEXT']
+        for ext in exts:
+            namext = name + ext
+            if os.path.isfile(namext):
+                return os.path.abspath(os.path.relpath(namext, cwd))
+    elif os.path.isfile(name) and name != os.path.basename(name):
+        return name
+    cc = builtins.__xonsh_commands_cache__
+    if name in cc:
+        # can be lazy here since we know name is already available
+        return cc.lazyget(name)[0]
+    else:
         return None
 
```

Comment:
I think there is an issue here. The commands are stored like: 'ex.exe' in cc, but the argument to `locate_binary` could have the wrong case and no extension. For example: `eX` and it should still find `ex.exe`. 

We could change the command cache to be case insensitive on windows. And then change the last part of `locate_binary()` to this: 
```python
    cc = builtins.__xonsh_commands_cache__
    if ON_WINDOWS:
        for ext in [''] + exts:
            nameext = (name + ext).lower()
            if nameext in (c.lower() for c in cc):
                return cc.lazyget(namext)[0]
    else:
        if name in cc:
            # can be lazy here since we know name is already available
            return cc.lazyget(name)[0]
    return None
```